### PR TITLE
fixes #25007 - fix medium providers for plugins

### PR DESCRIPTION
**Pull request by stbenjam**
Katello doesn't have medium set, since the medium is being provided
through the medium provider.  This makes @medium_provider not load
and result in empty `@kernel` and `@initrd` in pxe templates.

This changes the approach so that the default medium provider is the one that checks if medium is blank.  Also adds tests.

### CODE REVIEW

**Comment by stbenjam** [comment]
@lzap Do you think this is an ok alternative for discovery?

**Review by timogoebel — DISMISSED** [review]
Makes sense.

**Comment by timogoebel** [comment]
[test upgrade]

**Comment by timogoebel** [comment]
@stbenjam: Can you please take a look at the test failures? The random one i just checked looked related.

**Review by lzap — CHANGES_REQUESTED** [review]
Well this breaks discovery. Is it possible to keep `respond_to` call?

**Comment by stbenjam** [comment]
I think this version makes everyone happy. Can you all have another look?

Also: do I need to fix the codeclimate issue? I don't buy that decomposing the method into several reduces cognitive complexity.

**Comment by timogoebel** [comment]
> Also: do I need to fix the codeclimate issue? I don't buy that decomposing the method into several reduces cognitive complexity.

It's just meant as a hint for reviewers.

File: app/registries/foreman/plugin/medium_providers_registry.rb [hunk 1]
```diff
@@ -18,24 +19,28 @@ def register(provider)
...
       def find_provider(entity)
         raise Foreman::Exception.new('Must supply an entity to find a medium provider') unless entity
 
         provider_instances = providers.map { |provider| provider.constantize.new(entity) }
         valid_providers = provider_instances.select { |provider| provider.valid? }
         if valid_providers.count > 1
-          raise Foreman::Exception.new(
-            'Found more than one provider for %{entity}. Valid providers: %{providers}',
+          logger.error(
+            _('Found more than one provider for %{entity}. Valid providers: %{providers}') %
```
Comment (lzap) [inline app/registries/foreman/plugin/medium_providers_registry.rb]: We don't usually translate logs :)

File: app/registries/foreman/plugin/medium_providers_registry.rb [hunk 2]
```diff
@@ -18,24 +19,28 @@ def register(provider)
...
               entity: entity,
               providers: providers.map { |provider| provider.class.friendly_name }
             })
         end
 
         unless valid_providers.present?
-          raise Foreman::Exception.new(
-            'Could not find a provider for %{entity}. Providers returned %{errors}',
+          logger.error(
+            _('Could not find a provider for %{entity}. Providers returned %{errors}') %
```
Comment (lzap) [inline app/registries/foreman/plugin/medium_providers_registry.rb]: Ditto.

File: test/unit/medium_providers/plugin_test.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    end
+  end
+
+  setup do
+    Foreman::Plugin.medium_providers.register(PluginMediumProvider)
+  end
+
+  teardown do
+    Foreman::Plugin.medium_providers.unregister(PluginMediumProvider)
+  end
```
Comment (lzap) [inline test/unit/medium_providers/plugin_test.rb:20]: Does `teardown` get called even in case of exception? I hope so.

**Review by lzap — DISMISSED** [review]
I am hapy, rebase needed GH reports conflicts.

File: test/unit/medium_providers/plugin_test.rb (hunk 3 above)
Comment (stbenjam) [reply test/unit/medium_providers/plugin_test.rb:20]: It does

**Comment by stbenjam** [comment]
[test foreman]

**Review by timogoebel — APPROVED** [review]
Looks like an excellent patch to me.

@tbrisker: Are we going to include this in 1.20? I guess we want to be nice to the Katello folks here, don't we?

File: app/registries/foreman/plugin/medium_providers_registry.rb [hunk 4]
```diff
@@ -18,24 +19,28 @@ def register(provider)
...
-            'Found more than one provider for %{entity}. Valid providers: %{providers}',
+          logger.error(
+            'Found more than one provider for %{entity}. Valid providers: %{providers}' %
             {
               entity: entity,
               providers: providers.map { |provider| provider.class.friendly_name }
             })
         end
 
         unless valid_providers.present?
```
Comment (timogoebel) [inline app/registries/foreman/plugin/medium_providers_registry.rb:41]: You could return early here if you want.

`return valid_providers.first if valid_providers.present?`

Comment (stbenjam) [reply app/registries/foreman/plugin/medium_providers_registry.rb:41]: `valid_providers.first` will return nil if nothing's found, so I honestly prefer the obvious return at the end, unless you'd really prefer it that way

**Comment by stbenjam** [comment]
I would greatly appreciate it if provisioning worked with Foreman 1.20 and Katello :smile:

**Review by tbrisker — APPROVED** [review]
I'm fine with having this in 1.20, thanks @stbenjam, @lzap and @timogoebel!

**Pull request merged**